ibtk/include/ibtk/LibMeshSystemVectors.h

```diff
@@ -0,0 +1,185 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (c) 2020 - 2020 by the IBAMR developers
+// All rights reserved.
+//
+// This file is part of IBAMR.
+//
+// IBAMR is free software and is distributed under the 3-clause BSD
+// license. The full text of the license can be found in the file
+// COPYRIGHT at the top level directory of IBAMR.
+//
+// ---------------------------------------------------------------------
+
+#ifndef included_IBTK_LibMeshSystemVectors
+#define included_IBTK_LibMeshSystemVectors
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+
+#include <ibtk/FEDataManager.h>
+
+#include <libmesh/equation_systems.h>
+#include <libmesh/petsc_vector.h>
+#include <libmesh/system.h>
+
+#include <map>
+#include <memory>
+#include <string>
+#include <vector>
+
+/////////////////////////////// CLASS DEFINITION /////////////////////////////
+
+namespace IBTK
+{
+/*!
+ * \brief Class LibMeshSystemVectors is a convenience class that manages
+ * access to libMesh vectors for the same system defined on multiple
+ * parts. This class supports access to vectors ghosted with the Lagrangian
+ * partitioning (i.e., libMesh's computed partitioning) as well as the IB
+ * partitioning (i.e., the partitioning based on the distribution of SAMRAI
+ * data).
+ *
+ * This class stores information that depends on SAMRAI's parallel
+ * partitioning: LibMeshSystemVectors::reinit() should be called after
+ * regridding to clear this data.
+ *
+ * @note This class is intended for internal use in IBAMR. It is used in
+ * IBFEMethod and IBFESurfaceMethod.
+ */
+class LibMeshSystemVectors
+{
+public:
+    /// Constructor.
+    LibMeshSystemVectors(const std::vector<IBTK::FEDataManager*> fe_data_managers, std::string system_name)
+        : d_fe_data_managers(fe_data_managers), d_system_name(std::move(system_name))
+    {
+        for (IBTK::FEDataManager* fe_data_manager : d_fe_data_managers)
+        {
+            libMesh::EquationSystems* es = fe_data_manager->getEquationSystems();
+            d_systems.push_back(&es->get_system(d_system_name));
+        }
+    }
+
+    /*!
+     * Get a specific vector for a specific part.
+     */
+    libMesh::PetscVector<double>& getVector(const std::string& vec_name, const unsigned int part)
+    {
+        TBOX_ASSERT(part < d_fe_data_managers.size());
+        if (!vec_stored_in_map(vec_name))
+        {
+            if (vec_name == "solution") return dynamic_cast<libMesh::PetscVector<double>&>(*d_systems[part]->solution);
+            if (vec_name == "current local solution")
+                return dynamic_cast<libMesh::PetscVector<double>&>(*d_systems[part]->current_local_solution);
```

Comment:
Note that we hard-code some special vector names here since libMesh doesn't put these in its map of vectors - `system->solution` is not the same as `system->request_vector("solution")` but for convenience we are slightly inconsistent with libMesh here and identify every vector with a string.